File: app/src/main/java/com/example/jsontest/posts/PostsActivity.java
package com.example.jsontest.posts;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import com.example.jsontest.R;
import com.example.jsontest.ViewModelFactory;
import com.example.jsontest.comments.CommentsActivity;

public class PostsActivity extends AppCompatActivity {

    private RecyclerView mRecycler;
    private PostsAdapter mAdapter;

    private PostsViewModel mViewModel;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_list);

        mViewModel = ViewModelProviders.of(this, ViewModelFactory.getInstance(getApplication()))
                .get(PostsViewModel.class);

        mRecycler = findViewById(R.id.list_recycler);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        mRecycler.setLayoutManager(layoutManager);
        mAdapter = new PostsAdapter(mViewModel);
        mRecycler.setAdapter(mAdapter);

        mViewModel.postList.observe(this, list -> mAdapter.setData(list));

        mViewModel.clickedPost.observe(this, post -> {
            Intent intent = new Intent(this, CommentsActivity.class);
            intent.putExtra(CommentsActivity.EXTRA_POST, post);
            startActivity(intent);
        });
    }

    @Override
    protected void onStart() {
        super.onStart();
        mViewModel.start();
    }
}

File: app/src/main/java/com/example/jsontest/posts/PostsActivity.kt
package com.example.jsontest.posts

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.jsontest.R
import com.example.jsontest.ViewModelFactory
import com.example.jsontest.comments.CommentsActivity

class PostsActivity : AppCompatActivity() {

    private var mRecycler: RecyclerView? = null
    private var mAdapter: PostsAdapter? = null

    private var mViewModel: PostsViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_list)

        mViewModel = ViewModelProviders.of(this, ViewModelFactory(this))
            .get(PostsViewModel::class.java)

        mRecycler = findViewById(R.id.list_recycler)
        val layoutManager = LinearLayoutManager(this)
        layoutManager.orientation = RecyclerView.VERTICAL
        mRecycler!!.layoutManager = layoutManager
        mAdapter = PostsAdapter(mViewModel)
        mRecycler!!.adapter = mAdapter

        mViewModel!!.postList.observe(this, Observer { list -> mAdapter!!.setData(list) })

        mViewModel!!.clickedPost.observe(this, Observer { post ->
            val intent = Intent(this, CommentsActivity::class.java)
            intent.putExtra(CommentsActivity.EXTRA_POST, post)
            startActivity(intent)
        })
    }

    override fun onStart() {
        super.onStart()
        mViewModel!!.start()
    }
}
